### ISSUE TITLE
twisted.test.test_process.UtilTestCase should restore the rights of files it creates

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3202]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#3202|
|Type|enhancement|
|Created|2008-04-23 08:22:39Z|
|Branch|https://github.com/twisted/twisted/tree/test_process-perms-3202|

In particular, one of the "executable" file is read-only under windows, so that creates some problems. We can also delete the tree of files.

Attachments:

* [3202.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8d2/8d2cc24ba66f7f2d7e057609620538b8f96e2fee/e0fb40f31567fbae91ba2a2386d7d0fb8b1c0976.diff) (712 bytes) - added by therve on 2008-04-23 08:23:46Z - 
* [3202_2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8d2/8d2cc24ba66f7f2d7e057609620538b8f96e2fee/f983bc58bb0edef51e9ae937bf473df7f984793d.diff) (607 bytes) - added by therve on 2008-04-24 19:53:16Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3202 3202
type__enhancement enhancement
reporter__therve therve
priority__low low
milestone__ 
branch__branches_test_process_perms_3202 branches/test_process-perms-3202
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1208938959000000 1208938959000000
changetime__1209564986000000 1209564986000000
version__None None
owner__ 
cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve removed owner|
|-|-|

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set owner to @therve|
|-|-|

Humm.  I don't understand the motivation.  Can you elaborate a bit?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve removed owner|
|-|-|

In particular, it prevents trial from cleaning _trial_temp. Running trial successively in the same directory leaves plenty of old _trial_temp (because the read-only file can't be removed without a user interaction). It also crash combinator mkbranch on my machine.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @therve|
|-|-|

trial seems to also use `shutil.rmtree` to get rid of old `_trial_temp` directories.  I think I still don't understand, because it seems like the call to `rmtree` in `test_process` is just duplication of what trial is doing already.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

Hum good point. I'll investigate further.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve removed owner|
|-|-|

OK you were right, I don't know what I thought. The latest patch corrects the problem though.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @therve|
|-|-|

Looks good.  Please apply.

Maybe we can also think about teaching trial to be super awesome at deleting things so individual tests don't have to think about it as much.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set status to `closed`|
|-|-|

(In [23444]) Fix twisted.test.test_process.UtilTestCase so that it doesn't leave read-only
files in _trial_temp anymore.

Author: therve
Reviewer: exarkun
Fixes [#3202](https://github.com/twisted/twisted/issues/3202)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `reopened`|
|-|-|

(In [23447]) Revert r23444 - test suite regression on FreeBSD

Reopens: [#3202](https://github.com/twisted/twisted/issues/3202)

This changeset introduced three test failures on FreeBSD by way of an exception
in the modified tearDown method:

```
===============================================================================
[ERROR]: twisted.test.test_process.UtilTestCase.testWhich

Traceback (most recent call last):
  File "/usr/home/therve/Run/slave/freebsd-py2.4-select/Twisted/twisted/test/test_process.py", line 1963, in tearDown
    os.chmod(os.path.join(self.bazbar, "executable"), 700)
exceptions.OSError: [Errno 79] Inappropriate file type or format: 'twisted.test.test_process/UtilTestCase/testWhich/lo3a1X/temp/baz/bar/executable'
===============================================================================
[ERROR]: twisted.test.test_process.UtilTestCase.testWhichPathExt

Traceback (most recent call last):
  File "/usr/home/therve/Run/slave/freebsd-py2.4-select/Twisted/twisted/test/test_process.py", line 1963, in tearDown
    os.chmod(os.path.join(self.bazbar, "executable"), 700)
exceptions.OSError: [Errno 79] Inappropriate file type or format: 'twisted.test.test_process/UtilTestCase/testWhichPathExt/VchWFn/temp/baz/bar/executable'
===============================================================================
[ERROR]: twisted.test.test_process.UtilTestCase.test_whichWithoutPATH

Traceback (most recent call last):
  File "/usr/home/therve/Run/slave/freebsd-py2.4-select/Twisted/twisted/test/test_process.py", line 1963, in tearDown
    os.chmod(os.path.join(self.bazbar, "executable"), 700)
exceptions.OSError: [Errno 79] Inappropriate file type or format: 'twisted.test.test_process/UtilTestCase/test_whichWithoutPATH/OHS-2j/temp/baz/bar/executable'
-------------------------------------------------------------------------------
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [23448]) Branching to 'test_process-perms-3202'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `new`|
|-|-|

Sorry I am a bad reviewer.  I fixed the intent of the original patch in the branch, but now I'm not sure I understand the intent again. :)  If I had a Windows machine I might be able to figure it out, but since I don't, can you give me a hint?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve commented|
|-|-|

I think I've explained it [comment:3 here]:
> In particular, it prevents trial from cleaning _trial_temp. Running trial successively in the same directory leaves plenty of old _trial_temp (because the read-only file can't be removed without a user interaction).

The failed function is shutil.rmtree in _setUpTestdir. This case was mainly obvious on windows.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

I'm probably not understanding something about Windows.  Do you need write permission on a file to delete it on Windows?  On Linux, you only need write permission on the directory it is in.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve commented|
|-|-|

Yes, you need write permission to delete a file on Windows.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun removed owner|
|-|-|

Okay.  Amazing.  Now the change makes sense.  I adjusted `tearDown` to use `S_IWUSR` (and not set the file readable or executable as neither of those seems to be necessary for deletion) instead of an octal since that's a bit easier to read.  Otherwise the change is basically the same as the original.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_17"></a>|@therve set owner to @exarkun|
|-|-|

Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun set status to `closed`|
|-|-|

(In [23467]) Merge test_process-perms-3202

Author: exarkun
Reviewer: therve
Fixes: [#3202](https://github.com/twisted/twisted/issues/3202)

Restore write permission on a temporary directory created by the
tests so that files in it can be deleted by trial when the next
test run begins.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
